File: common/src/main/java/com/ssblur/scriptor/item/WritableSpellbook.java
package com.ssblur.scriptor.item;

import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.screens.inventory.BookEditScreen;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResultHolder;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.WritableBookItem;
import net.minecraft.world.level.Level;

public class WritableSpellbook extends WritableBookItem {
  public WritableSpellbook(Properties properties) {
    super(properties);
  }

  @Override
  public InteractionResultHolder<ItemStack> use(Level level, Player player, InteractionHand interactionHand) {
    var itemStack = player.getItemInHand(interactionHand);
    if(level.isClientSide)
      openScreen(player, itemStack, interactionHand);
    return InteractionResultHolder.sidedSuccess(itemStack, level.isClientSide());
  }

  @Environment(EnvType.CLIENT)
  public void openScreen(Player player, ItemStack itemStack, InteractionHand interactionHand) {
    Minecraft.getInstance().setScreen(new BookEditScreen(player, itemStack, interactionHand));
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/item/books/WritableSpellbook.kt
package com.ssblur.scriptor.item.books

import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.screens.inventory.BookEditScreen
import net.minecraft.world.InteractionHand
import net.minecraft.world.InteractionResultHolder
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.WritableBookItem
import net.minecraft.world.level.Level

class WritableSpellbook(properties: Properties) : WritableBookItem(properties) {
    override fun use(
        level: Level,
        player: Player,
        interactionHand: InteractionHand
    ): InteractionResultHolder<ItemStack> {
        val itemStack = player.getItemInHand(interactionHand)
        if (level.isClientSide) openScreen(player, itemStack, interactionHand)
        return InteractionResultHolder.sidedSuccess(itemStack, level.isClientSide())
    }

    @Environment(EnvType.CLIENT)
    fun openScreen(player: Player, itemStack: ItemStack, interactionHand: InteractionHand) {
        Minecraft.getInstance().setScreen(BookEditScreen(player, itemStack, interactionHand))
    }
}
